### PR TITLE
Fix extraction of product bundle ID for iOS projects

### PR DESCRIPTION
**Pull request by mravn-google**
#21216 broke extraction of a product bundle ID from iOS projects.
#21225 applied a quick fix to a down-stream error.

This PR addresses the root cause.

### CODE REVIEW

File: packages/flutter_tools/lib/src/application_package.dart [hunk 1]
```diff
@@ -34,7 +34,7 @@ abstract class ApplicationPackage {
   File get packagesFile => null;
 
   @override
-  String toString() => displayName;
+  String toString() => displayName ?? 'application';
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/application_package.dart]: Why can this ever be null?

File: packages/flutter_tools/lib/src/project.dart [hunk 2]
```diff
@@ -174,22 +177,47 @@ class IosProject {
...
   /// The '.pbxproj' file of the host app.
   File get xcodeProjectInfoFile => xcodeProject.childFile('project.pbxproj');
 
-  /// The product bundle identifier of the host app.
+  /// The product bundle identifier of the host app, or null if not set or if
+  /// iOS tooling needed to read it is not installed.
   String get productBundleIdentifier {
-    return _firstMatchInFile(xcodeProjectInfoFile, _productBundleIdPattern)?.group(1);
+    final String fromPbxproj = _firstMatchInFile(xcodeProjectInfoFile, _productBundleIdPattern)?.group(1);
+    final String fromPlist = iosWorkflow.getPlistValueFromFile(
```
Comment (sigurdm) [inline packages/flutter_tools/lib/src/project.dart]: Perhaps check this first to avoid looking in pbxproj if not needed

**Review by sigurdm — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/application_package.dart (hunk 1 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/application_package.dart]: Replaced `'application'` by `id` which is asserted not to be null. This avoids silent errors. Thanks.

File: packages/flutter_tools/lib/src/project.dart (hunk 2 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/project.dart]: Done.

**Review by mravn-google — COMMENTED** [review]
Thanks for the review.

**Pull request merged**